File: src/main/java/com/github/morningzeng/toolset/utils/LanguageUtils.java
package com.github.morningzeng.toolset.utils;

import com.intellij.json.json5.Json5Language;
import com.intellij.lang.Language;
import com.intellij.lang.html.HTMLLanguage;
import com.intellij.lang.xml.XMLLanguage;
import com.intellij.openapi.fileTypes.PlainTextLanguage;

import java.util.List;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author Morning Zeng
 * @since 2024-06-20
 */
public final class LanguageUtils {

    public static Language tryResolve(String content) {
        content = content.trim();
        final List<String> list = content.lines().toList();
        if (list.isEmpty()) {
            return PlainTextLanguage.INSTANCE;
        }
        final String neckLine = list.get(0);
        final String shadow = list.get(list.size() - 1);
        if (neckLine.startsWith("<html>") && shadow.endsWith("</html>")) {
            return HTMLLanguage.INSTANCE;
        }
        if ((neckLine.startsWith("{") && shadow.endsWith("}"))
                || (neckLine.startsWith("[") && shadow.endsWith("]"))) {
            return Json5Language.INSTANCE;
        }
        final Pattern startXml = Pattern.compile("^(?<=<).+?(?=>)");
        final Pattern endXml = Pattern.compile("(?<=</).+?(?=>)$");
        final Matcher startXmlMatcher = startXml.matcher(neckLine);
        final Matcher endXmlMatcher = endXml.matcher(shadow);
        if (startXmlMatcher.find() && endXmlMatcher.find()) {
            if (Objects.equals(startXmlMatcher.group(), endXmlMatcher.group())) {
                return XMLLanguage.INSTANCE;
            }
        }
        return PlainTextLanguage.INSTANCE;
    }

}

File: src/main/kotlin/com/github/morningzeng/toolbox/utils/LanguageUtils.kt
package com.github.morningzeng.toolbox.utils

import com.github.morningzeng.toolbox.ui.component.LanguageTextArea
import com.intellij.json.json5.Json5Language
import com.intellij.lang.html.HTMLLanguage
import com.intellij.lang.xml.XMLLanguage
import com.intellij.openapi.fileTypes.PlainTextLanguage

/**
 * @author Morning Zeng
 * @since 2025-08-15
 */
object LanguageUtils {

    fun LanguageTextArea.detectLanguage() {
        this.language = when {
            isJsonContent(this.text) -> Json5Language.INSTANCE
            isXmlContent(this.text) -> XMLLanguage.INSTANCE
            isHtmlContent(this.text) -> HTMLLanguage.INSTANCE
            else -> PlainTextLanguage.INSTANCE
        }
    }

    private fun isJsonContent(content: String): Boolean {
        val trimmed = content.trim()
        return (trimmed.startsWith("{") && trimmed.endsWith("}")) ||
                (trimmed.startsWith("[") && trimmed.endsWith("]"))
    }

    private fun isXmlContent(content: String): Boolean {
        return content.contains("<?xml") ||
                (content.contains("<") && content.contains(">") && content.contains("</"))
    }

    private fun isHtmlContent(content: String): Boolean {
        val htmlTags = arrayOf(
            "<!DOCTYPE", "<html", "<head", "<body", "<div", "<span", "<p>", "<h1", "<h2"
        )
        val lowerContent = content.lowercase()
        return htmlTags.any { lowerContent.contains(it.lowercase()) }
    }


}